### ISSUE TITLE
DaemonSet to support "add first, then delete" rolling update 

### ISSUE THREAD

**Issue by janetkuo** [kubernetes/kubernetes#48841]
DaemonSet currently supports "delete first, then add" rolling update strategy.

This is a feature request for DaemonSet to support "add first, then delete" rolling update. This can be done with a new upgrade strategy. Users will use `maxSurge` in addition to `maxUnavailable` in this new upgrade strategy. With this, each node can have at least one Daemon pod running on it during a rolling update. 

cc @kubernetes/sig-apps-feature-requests

**Comment by liqlin2015** [comment]
See this doc https://kubernetes.io/docs/tasks/manage-daemon/update-daemon-set/

**Comment by kargakis** [comment]
> This can be done with a new upgrade strategy.

Can't it be added in the existing RollingUpdate strategy? Why?

**Comment by lukaszo** [comment]
+1 for changing RollingUpdate strategy. If maxSurge defaults to 0 it will be backward compatible.

**Comment by luxas** [comment]
> Can't it be added in the existing RollingUpdate strategy? Why?

I guess it would be possible with the current strategy. If MaxUnavailable is 0 I guess the end result is that the controller will add a new Pod before removing the old one.

Do we even need to add a MaxSurge option for this to work?

**Comment by kargakis** [comment]
> Do we even need to add a MaxSurge option for this to work?

Any alterantives?

**Comment by janetkuo** [comment]
> Do we even need to add a MaxSurge option for this to work?

This is required for DaemonSet controller to know how many more replicas the users allow it to create.

> Can't it be added in the existing RollingUpdate strategy? Why?

The reason that we need a new strategy is that current RollingUpdate strategy requires `MaxUnavailable > 0`. Changing that requirement wouldn't be backward compatible. We can implement this without a new strategy by changing that validation in a new API version, but the controller logic needs to be changed together in order to support it.

**Comment by timothysc** [comment]
So we need this for some of the self hosting work.  @janetkuo is this planned for 1.8?

**Comment by kargakis** [comment]
We also need to deprecate/remove TemplateGeneration so bumping to v1beta2 makes sense to me. Do we need to bump the rest of the apps group?

**Comment by janetkuo** [comment]
Bumping DaemonSet to apps/v1beta2 is planned for 1.8. Supporting this feature is not planned for 1.8.

> Do we need to bump the rest of the apps group?

By the end of 1.8, Deployment, ReplicaSet, StatefulSet, DaemonSet will be in apps/v1beta2

**Comment by luxas** [comment]
> Supporting this feature is not planned for 1.8

So this "add first, then delete" strategy will _not_ make v1.8, right?
That'd be unfortunate. We can work around it I _think_, but it'll be hacky for sure.

> By the end of 1.8, Deployment, ReplicaSet, StatefulSet, DaemonSet will be in apps/v1beta2

SGTM :+1:

**Comment by luxas** [comment]
> Adding this feature may bring some risk to our GA plan. If it's a big blocker for you. Please discuss this with sig apps team.

@janetkuo Thanks. @aaronlevy said he'd be able to put resources from CoreOS on it this cycle if sig-apps approved. We'd really want to get this, but I also understand the reasoning behind stability.

What do you think @kubernetes/sig-apps-feature-requests? Would this be possible to get into v1.8 if you'd get some folks dedicated to implementing it?

**Comment by kargakis** [comment]
I would be fine with having this, it brings DaemonSet RollingUpdates on par with Deployments and since we will release v1beta2 in 1.8, it's a chance to get it in the existing strategy as opposed to a separate one.

**Comment by kow3ns** [comment]
1. I would think we have to address [DaemonSet's lack of sensitivity to node resources] (#https://github.com/kubernetes/kubernetes/issues/46935) prior to considering this. 
1. This updateStrategy would probably only work for self hosting if we choose to [not go through the default scheduler](https://github.com/kubernetes/kubernetes/issues/42002) for Pod creation. Are we committed to not doing this ever, or do we have a notional design that would allow DaemonSet to host controller manager and go through the default scheduler?
1. I don't feel strongly about it, but wouldn't it be cleaner (and lower risk) to implement this as a separate UpdateStrategy?
1. I think we can implement this, in a backward compatible way, after settling 1 and 2, and after stabilizing and advancing the API to GA.

**Comment by resouer** [comment]
+1 for `I think we can implement this, in a backward compatible way, after settling 1 and 2, and after stabilizing and advancing the API to GA.`

**Comment by lukaszo** [comment]
> I don't feel strongly about it, but wouldn't it be cleaner (and lower risk) to implement this as a separate UpdateStrategy?

For deployments it's one strategy. I think for all controllers we should make strategies as similar as possible.

**Comment by roberthbailey** [comment]
> we want people to have used the rolling update feature for 2 quarters and gave feedback before we bring DaemonSet to GA.

We *want* to use DaemonSet rolling updates but it doesn't currently meet our needs. I feel like that is feedback that should be taken into account as we promote this API towards GA.

**Comment by roberthbailey** [comment]
@kow3ns - 

> This updateStrategy would probably only work for self hosting if we choose to not go through the default scheduler for Pod creation.

I'm not sure I understand. If we "create then delete" for the controller manager and the scheduler, then we will have two copies running before we delete the old one. So in both cases there won't be a time window during which there is no binary that keeps the control plane functional.

**Comment by luxas** [comment]
> For deployments it's one strategy. I think for all controllers we should make strategies as similar as possible.

I agree with @lukaszo, API predictability/consistency is important. To me it seems like making this change at the same time as you do the shift to v1beta2 is the best move, regardless of whether that happens in v1.8 or not. The validation for MaxSurge/MaxUnavailable will just be different in v1beta2, and allow `MaxUnavailable=0` for instance. This is not allowed if you use the `extensions/v1beta1` object. In the controller loops you'll just handle the `MaxUnavailable=0` case, as you're always dealing with the internal version of the object.
Makes sense?

Also want to reiterate that @aaronlevy offered to help get this done.

**Comment by janetkuo** [comment]
@luxas @aaronlevy @roberthbailey would you join the next [sig-apps meeting](https://github.com/kubernetes/community/tree/master/sig-apps#meetings) (July 24, 2017 16:00 UTC) to discuss this? We'll do 1.8 planning (see [agenda](https://docs.google.com/document/d/1LZLBGW2wRDwAfdBNHJjFfk9CFoyZPcIYGWU7R1PQ3ng/edit#heading=h.fdccksqewtn6)).

**Comment by luxas** [comment]
@janetkuo Yes, I'll join.

@timothysc ^

**Comment by aaronlevy** [comment]
To follow up, @diegs is interested in taking a look at this.

I wasn't able to attend most of the meeting so I could be off here, but it sounded like a goal might be to try and keep the code for the solution somewhat separate from the existing policy to reduce the risk that this could introduce.

So would this mean a different strategy would be preferable to the api bump + MaxUnavailable=0 behavior change?

**Comment by kargakis** [comment]
The api bump will happen either way as part of bringing the controllers closer to GA. Initially, I advocated for adding the new option in the existing strategy but that is riskier, and until more use-cases pop up, a separate strategy is preferrable.

**Comment by fasaxc** [comment]
I just ran into this while doing rolling update testing of our networking plugin.  Starting the new pod first would confuse our daemon since it only expects one instance.  Is there any mileage in enhancing the kill-first algorithm to pre-pull the new image and maybe pre-prep the new pod but to still do the cutover as kill-then-start.  That could avoid the headaches of temporarily doubling CPU usage while giving most of the benefit.

Maybe a field on the new podspec that says "block until <other pod> is killed" so kubelet knows to prep the pod and pull the image but waits until the other pod is actually dead before starting it?

**Comment by lukaszo** [comment]
@fasaxc Currently `RollingUpdate` strategy first kills the old pod and then starts a new one.
If we add `MaxSurge` option to the strategy it will be configurable.

**Comment by fasaxc** [comment]
@lukaszo Yeah, I understand what was being proposed in this issue.  I was wondering if the kill-then-start mode can be improved to reduce the downtime.  (Start-then-kill requires apps to be updated to support running multiple instances in the transition period and it sounds like there are problems to be overcome in quota/scheduling.)

**Comment by ant31** [comment]
@fasaxc could you open another issue to discuss your proposition ? 

@janetkuo is there a decision about integrating this feature before moving GA?

**Comment by luxas** [comment]
@ant31 The decision is to proceed with a new start-then-kill UpdateStrategy and add that feature to v1.8
@diegs (mentored by @aaronlevy and others) will primarily do the coding for this feature and sig-apps will review and approve.

@roberthbailey @janetkuo @erictune @aaronlevy I guess we should make a features issue for this, can anyone of you do that or approve to create one?

**Comment by roberthbailey** [comment]
Feature issue is https://github.com/kubernetes/features/issues/373

**Comment by diegs** [comment]
First cut of implementation is in #51161, cc @luxas

**Comment by kow3ns** [comment]
is there any design for this feature?

**Comment by erictune** [comment]
+1 to design before code.  

Specific questions for design to address:
  - What is a concrete example application (prefer more than 1) that will use this upgrade strategy
  - If  old pod and the new pod need to use the same port, how is control of the port transfered.  There are several ways I can imagine doing this.  Which one(s) do we want to support/endorse.
  - Are there other resources besides ports that need to be transfered during update?  E.g. use of a local device?  How is this mediated?  
  - Call out things that don't work: like if each pod needs 1 GPU resource, they can't share it.  Or if they both need a particular type of core, they can't do this.  Interacts with Numa proposal.
  - Verify that the scheduler will allow two pods that need the same hostPort to schedule on the same machine at all.

**Comment by diegs** [comment]
Thanks for the early reviews. Added a design doc (included your questions @erictune):

https://github.com/kubernetes/community/pull/977

Happy to continue discussion there before reviewing this PR.

**Comment by luxas** [comment]
This was abandoned for now as per the proposal discussion. Closing

**Comment by janetkuo** [comment]
Here's a KEP for adding MaxSurge to DaemonSet https://github.com/kubernetes/enhancements/pull/1590

**Comment by smarterclayton** [comment]
This went alpha in 1.21 behind a feature gate, please give feedback so we can assess promotion.

**Comment by da-head0** [comment]
@smarterclayton 
so add first, then delete is now applied?
when I see newest kubernetes document, It says

> With RollingUpdate update strategy, after you update a DaemonSet template, old DaemonSet pods will be killed, and new DaemonSet pods will be created automatically,
